File: storage.py
"""Storage implementations for the application"""
import time
from datetime import datetime
from typing import Dict, List, Optional, Any

class MemStorage:
    """In-memory storage implementation for Flask backend"""
    
    def __init__(self):
        """Initialize the storage with empty collections"""
        self.users = {}  # user_id -> user data
        self.projects = {}  # project_id -> project data
        self.brand_concepts = {}  # concept_id -> concept data
        
        self.user_id_counter = 1
        self.project_id_counter = 1
        self.concept_id_counter = 1
        
        # Create a default user for demo purposes
        self.create_user({
            "username": "demo",
            "password": "password"  # In a real app, this would be hashed
        })
        
        # Add a sample project
        self.create_project({
            "name": "Solystra",
            "clientName": "Sample Client",
            "userId": 1
        })

    def get_user(self, id: int) -> Optional[Dict]:
        """Get user by ID"""
        return self.users.get(id)
    
    def get_user_by_username(self, username: str) -> Optional[Dict]:
        """Get user by username"""
        for user in self.users.values():
            if user["username"] == username:
                return user
        return None
    
    def create_user(self, user_data: Dict) -> Dict:
        """Create a new user"""
        user_id = self.user_id_counter
        self.user_id_counter += 1
        
        user = {
            "id": user_id,
            **user_data
        }
        
        self.users[user_id] = user
        return user
    
    def get_projects(self, user_id: int) -> List[Dict]:
        """Get projects for a user"""
        return [p for p in self.projects.values() if p["userId"] == user_id]
    
    def get_project(self, id: int) -> Optional[Dict]:
        """Get project by ID"""
        return self.projects.get(id)
    
    def create_project(self, project_data: Dict) -> Dict:
        """Create a new project"""
        project_id = self.project_id_counter
        self.project_id_counter += 1
        
        project = {
            "id": project_id,
            "createdAt": datetime.now().isoformat(),
            **project_data
        }
        
        self.projects[project_id] = project
        return project
    
    def update_project(self, id: int, partial_project: Dict) -> Optional[Dict]:
        """Update an existing project"""
        project = self.get_project(id)
        if not project:
            return None
        
        updated_project = {**project, **partial_project}
        self.projects[id] = updated_project
        return updated_project
    
    def delete_project(self, id: int) -> bool:
        """Delete a project and its concepts"""
        if id not in self.projects:
            return False
        
        # Delete all brand concepts for this project
        concept_ids_to_delete = []
        for concept_id, concept in self.brand_concepts.items():
            if concept["projectId"] == id:
                concept_ids_to_delete.append(concept_id)
        
        for concept_id in concept_ids_to_delete:
            del self.brand_concepts[concept_id]
        
        # Delete the project
        del self.projects[id]
        return True
    
    def get_brand_concepts(self, project_id: int) -> List[Dict]:
        """Get all brand concepts for a project"""
        return [c for c in self.brand_concepts.values() if c["projectId"] == project_id]
    
    def get_brand_concept(self, id: int) -> Optional[Dict]:
        """Get brand concept by ID"""
        return self.brand_concepts.get(id)
    
    def create_brand_concept(self, concept_data: Dict) -> Dict:
        """Create a new brand concept"""
        concept_id = self.concept_id_counter
        self.concept_id_counter += 1
        
        # Set isActive to False by default
        if "isActive" not in concept_data:
            concept_data["isActive"] = False
        
        concept = {
            "id": concept_id,
            "createdAt": datetime.now().isoformat(),
            **concept_data
        }
        
        self.brand_concepts[concept_id] = concept
        
        # If this is the first concept for the project, make it active
        project_concepts = self.get_brand_concepts(concept["projectId"])
        if len(project_concepts) == 1:
            concept["isActive"] = True
            self.brand_concepts[concept_id] = concept
        
        return concept
    
    def update_brand_concept(self, id: int, partial_concept: Dict) -> Optional[Dict]:
        """Update an existing brand concept"""
        concept = self.get_brand_concept(id)
        if not concept:
            return None
        
        updated_concept = {**concept, **partial_concept}
        self.brand_concepts[id] = updated_concept
        return updated_concept
    
    def delete_brand_concept(self, id: int) -> bool:
        """Delete a brand concept"""
        if id not in self.brand_concepts:
            return False
        
        # Get the concept to check if it's active
        concept = self.brand_concepts[id]
        project_id = concept["projectId"]
        was_active = concept["isActive"]
        
        # Delete the concept
        del self.brand_concepts[id]
        
        # If the deleted concept was active, set another concept as active (if any remain)
        if was_active:
            project_concepts = self.get_brand_concepts(project_id)
            if project_concepts:
                # Set the first available concept as active
                first_concept_id = project_concepts[0]["id"]
                self.set_active_brand_concept(first_concept_id, project_id)
        
        return True
    
    def set_active_brand_concept(self, id: int, project_id: int) -> bool:
        """Set a brand concept as active and deactivate others for the same project"""
        # Check if the concept exists
        concept = self.get_brand_concept(id)
        if not concept:
            return False
        
        # Deactivate all concepts for this project
        for c_id, c in self.brand_concepts.items():
            if c["projectId"] == project_id:
                c["isActive"] = False
                self.brand_concepts[c_id] = c
        
        # Activate the specified concept
        concept["isActive"] = True
        self.brand_concepts[id] = concept
        
        return True

File: server/storage.ts
import type { 
  User, InsertUser, Project, InsertProject,
  BrandConcept, InsertBrandConcept,
  BrandInput, BrandOutput
} from "@shared/schema";

import {
  users, projects, brandConcepts
} from "@shared/schema";

import { postgresStorage, db } from './db';
import { eq, desc } from "drizzle-orm";

// Using direct Supabase PostgreSQL connection via pooled connection string

export interface IStorage {
  // User operations
  getUser(id: number): Promise<User | undefined>;
  getUserByUsername(username: string): Promise<User | undefined>;
  createUser(user: InsertUser): Promise<User>;
  
  // Project operations
  getProjects(userId: number): Promise<Project[]>;
  getProject(id: number): Promise<Project | undefined>;
  createProject(project: InsertProject): Promise<Project>;
  updateProject(id: number, project: Partial<Project>): Promise<Project | undefined>;
  deleteProject(id: number): Promise<boolean>;
  
  // Brand concept operations
  getBrandConcepts(projectId: number): Promise<BrandConcept[]>;
  getBrandConcept(id: number): Promise<BrandConcept | undefined>;
  createBrandConcept(concept: InsertBrandConcept): Promise<BrandConcept>;
  updateBrandConcept(id: number, concept: Partial<BrandConcept>): Promise<BrandConcept | undefined>;
  deleteBrandConcept(id: number): Promise<boolean>;
  setActiveBrandConcept(id: number, projectId: number): Promise<boolean>;
}export class DatabaseStorage implements IStorage {
  async getUser(id: number): Promise<User | undefined> {
    if (!db) {
      console.warn('Database not initialized. Cannot get user.');
      return undefined;
    }
    
    const [user] = await db.select().from(users).where(eq(users.id, id));
    return user || undefined;
  }

  async getUserByUsername(username: string): Promise<User | undefined> {
    if (!db) {
      console.warn('Database not initialized. Cannot get user by username.');
      return undefined;
    }
    
    const [user] = await db.select().from(users).where(eq(users.username, username));
    return user || undefined;
  }

  async createUser(insertUser: InsertUser): Promise<User> {
    if (!db) {
      console.warn('Database not initialized. Cannot create user.');
      throw new Error('Database not initialized');
    }
    
    const [user] = await db
      .insert(users)
      .values(insertUser)
      .returning();
    return user;
  }
  
  // Project operations
  async getProjects(userId: number): Promise<Project[]> {
    if (!db) {
      console.warn('Database not initialized. Cannot get projects.');
      return [];
    }
    
    const projectsTable = projects;
    const projectsData = await db
      .select()
      .from(projectsTable)
      .where(eq(projectsTable.userId, userId))
      .orderBy(desc(projectsTable.createdAt));
    return projectsData;
  }
  
  async getProject(id: number): Promise<Project | undefined> {
    if (!db) {
      console.warn('Database not initialized. Cannot get project.');
      return undefined;
    }
    
    const projectsTable = projects;
    const [project] = await db
      .select()
      .from(projectsTable)
      .where(eq(projectsTable.id, id));
    return project || undefined;
  }
  
  async createProject(project: InsertProject): Promise<Project> {
    if (!db) {
      console.warn('Database not initialized. Cannot create project.');
      throw new Error('Database not initialized');
    }
    
    const projectsTable = projects;
    const [newProject] = await db
      .insert(projectsTable)
      .values(project)
      .returning();
    return newProject;
  }
  
  async updateProject(id: number, project: Partial<Project>): Promise<Project | undefined> {
    if (!db) {
      console.warn('Database not initialized. Cannot update project.');
      return undefined;
    }
    
    // Exclude id from update
    const { id: _, ...updateValues } = project;
    
    if (Object.keys(updateValues).length === 0) {
      return this.getProject(id);
    }
    
    const projectsTable = projects;
    const [updatedProject] = await db
      .update(projectsTable)
      .set(updateValues)
      .where(eq(projectsTable.id, id))
      .returning();
    return updatedProject || undefined;
  }
  
  async deleteProject(id: number): Promise<boolean> {
    if (!db) {
      console.warn('Database not initialized. Cannot delete project.');
      return false;
    }
    
    const projectsTable = projects;
    const result = await db
      .delete(projectsTable)
      .where(eq(projectsTable.id, id))
      .returning();
    return result.length > 0;
  }
  
  // Brand concept operations
  async getBrandConcepts(projectId: number): Promise<BrandConcept[]> {
    if (!db) {
      console.warn('Database not initialized. Cannot get brand concepts.');
      return [];
    }
    
    const conceptsTable = brandConcepts;
    const concepts = await db
      .select()
      .from(conceptsTable)
      .where(eq(conceptsTable.projectId, projectId))
      .orderBy(desc(conceptsTable.createdAt));
    return concepts;
  }
  
  async getBrandConcept(id: number): Promise<BrandConcept | undefined> {
    if (!db) {
      console.warn('Database not initialized. Cannot get brand concept.');
      return undefined;
    }
    
    const conceptsTable = brandConcepts;
    const [concept] = await db
      .select()
      .from(conceptsTable)
      .where(eq(conceptsTable.id, id));
    return concept || undefined;
  }
  
  async createBrandConcept(concept: InsertBrandConcept): Promise<BrandConcept> {
    if (!db) {
      console.warn('Database not initialized. Cannot create brand concept.');
      throw new Error('Database not initialized');
    }
    
    const conceptsTable = brandConcepts;
    const [newConcept] = await db
      .insert(conceptsTable)
      .values(concept)
      .returning();
    
    // If this concept is active, deactivate others for this project
    if (concept.isActive) {
      await this.setActiveBrandConcept(newConcept.id, concept.projectId);
    }
    
    return newConcept;
  }
  
  async updateBrandConcept(id: number, concept: Partial<BrandConcept>): Promise<BrandConcept | undefined> {
    if (!db) {
      console.warn('Database not initialized. Cannot update brand concept.');
      return undefined;
    }
    
    // Store if isActive is true so we can update other concepts if needed
    const isBeingSetActive = concept.isActive === true;
    const projectId = concept.projectId;
    
    // Exclude id from update
    const { id: _, ...updateValues } = concept;
    
    if (Object.keys(updateValues).length === 0) {
      return this.getBrandConcept(id);
    }
    
    const conceptsTable = brandConcepts;
    const [updatedConcept] = await db
      .update(conceptsTable)
      .set(updateValues)
      .where(eq(conceptsTable.id, id))
      .returning();
    
    // If this concept is being set as active, deactivate others
    if (isBeingSetActive && projectId) {
      await this.setActiveBrandConcept(id, projectId);
    }
    
    return updatedConcept || undefined;
  }
  
  async deleteBrandConcept(id: number): Promise<boolean> {
    if (!db) {
      console.warn('Database not initialized. Cannot delete brand concept.');
      return false;
    }
    
    const conceptsTable = brandConcepts;
    const result = await db
      .delete(conceptsTable)
      .where(eq(conceptsTable.id, id))
      .returning();
    return result.length > 0;
  }
  
  async setActiveBrandConcept(id: number, projectId: number): Promise<boolean> {
    if (!db) {
      console.warn('Database not initialized. Cannot set active brand concept.');
      return false;
    }
    
    try {
      const conceptsTable = brandConcepts;
      // First, deactivate all concepts for this project
      await db
        .update(conceptsTable)
        .set({ isActive: false })
        .where(eq(conceptsTable.projectId, projectId));
      
      // Then, activate the specified concept
      const result = await db
        .update(conceptsTable)
        .set({ isActive: true })
        .where(eq(conceptsTable.id, id))
        .returning();
      
      return result.length > 0;
    } catch (error: any) {
      console.error('Error setting active brand concept:', error);
      return false;
    }
  }
}

// Use Supabase PostgreSQL connection via their pooled connection string
// Connection to Supabase database directly (bypassing the REST API)
// SUPABASE_DB_URL is already a pooled connection managed by Supabase
console.log('Using Supabase PostgreSQL via pooled connection string');

// Export PostgreSQL storage implementation
export const storage = postgresStorage;
